### ISSUE TITLE
`create_result` returns x with lowest observed f(x), not lowest predicted f(x)

### ISSUE THREAD

**Issue by fredcallaway** [scikit-optimize/scikit-optimize#505]
This seems wrong to me. The optimizer seems to be aware of its uncertainty, re-evaluating the "fluke point" and discovering that it's actually not usually that great. But then it gives me the fluke point at the end! This is especially problematic when variance is non-uniform and lower near the true minimum.

Ideally, the convergence plot would be updated as well, showing the expected minimum as well as the variance around that estimate. This would lead to the convergence plot being non-monotonic, but it would converge on the true minimum rather than some other point.

**Comment by glouppe** [comment]
Hi,

Maybe you are looking for `skopt.utils.expected_minimum`?

See https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/utils.py#L197

**Comment by fredcallaway** [comment]
Thanks, I'm embarrassed to admit I discovered that function after posting the issue.

Still, I think my point holds, for two reasons. First, it's not intuitive that you would have to use another function to get the expected best value. The expected value is what you really care about (right?) so this should determine the primary value returned, i.e. `result.x`. Second because `expected_minimum` is not constrained to points that have been evaluated, it can return an `x` from a relatively unexplored part of the space. Empirically, I've found it to perform worse than the best of the tested points:

```py
gp = res.models[-1]
res.x_iters[gp.predict(res.x_iters).argmin()]
```

**Comment by glouppe** [comment]
I am not convinced there is a "right" or "wrong" answer for that. 

i) At present, we return the minimum among the function evaluations. We have chosen to return that because this is the only truth we have access to. The GP is only a model of these evaluations and therefore can be wrong.

ii) `expected_minimum` is supposed to return the expected minimum among the predictions of the GP over the entire domain. It might indeed return a point that has not been tested.

iii) you suggest the return the expected  minimum among the predictions of the GP, over the points that have been tested. 

I think all three approaches are reasonable. Maybe we could have another function for iii) (or a flag in `expected_minimum`). What do you think?

**Comment by betatim** [comment]
I'd be -1 on changing what `create_results` returns. It would cause confusion about what is in `res.x` if we change it and the current approach has the benefit of being easy to understand. Hopefully people can quickly work out what actually is in `res.x` and then go "ahrg, this isn't what I wanted". Combine that with some better documentation on finding `expected_minimum` and the difficulty of deciding what the "right" answer is. IMHO that would be pretty good.

My proposal: write a short tutorial highlighting the three choices @glouppe outlined, when you might want which one, and how to get it.

---

For my education: do we have an example where `expected_minimum` picks a point in an unexplored area? I'd have expected that the minimization procedure would place at least some samples in the vicinity of the minimum of the model.

**Comment by fredcallaway** [comment]
OK, I'm convinced that `res.x` shouldn't change. I did some more experimentation, and found results conflicting with what I reported above—`res.x` now seems to perform better than `expected_minimum` (ii) or `expected_minimum(tested_points_only=False)` (iii). I think a more careful controlled experiment is in order, which could be written up as a tutorial. I'm happy to do this.

@betatim, I think your intuition is correct. This was the only explanation I could come up with for (iii) beating (ii). I was getting those results with few iterations on a very noisy objective function. So I can imagine that the final test point could dramatically change the GP and hence the expected minimum.

**Comment by betatim** [comment]
What do you think about writing a short notebook for the examples section that demonstrates the three different options for "the minimum"?

**Comment by iaroslav-ai** [comment]
I would be quite interested in this myself. I often train all kinds of neural networks, and there I often get quite noisy results with fixed parameters. So for my work to be reproducible, I would be rather interested in finding points where the  expected objective value (loss at termination of training algo) is best. However, for practical applications I am simply looking for a situation with best objective. So I think it is a great idea have a notebook as you describe highlighting this. I think this will help many people to make better use of skopt.
